### PR TITLE
Adding #time function

### PR DESCRIPTION
**Pull request by omarzd**
# time should return the number of milliseconds since the epoch. However,

the return type is long, and long isn't supported in K as far as I know.
I made this commit, even though it's a mess, to get feedback from the
developers on how to properly do this. I was following [this pull
request](https://github.com/kframework/k/pull/967/files) which is why I
put it under IO but then I changed my mind in hooks.properties and made
time.k!

### CODE REVIEW

**Comment by dwightguth** [comment]
Just make the return type of #time be an Int. Ints are arbitrary precision in K, so you don't have to worry about it being out of range.

File: k-distribution/include/builtins/time.k [hunk 1]
```diff
@@ -0,0 +1,10 @@
+// Copyright (c) 2012-2014 K Team. All Rights Reserved.
+require "long.k"
+
+module TIME-HOOKS
+  imports #TIME-INTERFACE
+  imports LONG-HOOKS
+
+  syntax Long ::= "time" "(" ")" [function, klabel(randomRandom), hook(#TIME:time)]
```
Comment (dwightguth) [inline k-distribution/include/builtins/time.k]: fix this klabel

File: k-distribution/include/io/util.k [hunk 2]
```diff
@@ -0,0 +1,4 @@
+// Copyright (c) 2013-2014 K Team. All Rights Reserved.
```
Comment (dwightguth) [inline k-distribution/include/io/util.k]: should this file be deleted?

File: k-distribution/include/builtins/time.k [hunk 3]
```diff
@@ -0,0 +1,10 @@
+// Copyright (c) 2012-2014 K Team. All Rights Reserved.
```
Comment (dwightguth) [inline k-distribution/include/builtins/time.k]: this is a new file so the copyright should just be 2014

**Comment by dwightguth** [comment]
There is one more component of this that you need to handle in order to make it work with maude, which is the actual invocation of the IO server. Take a look at some code in tcp.k for examples of how the IO server is called.

**Comment by dwightguth** [comment]
One thing I'd like to note is that we actually plan to deprecate the Maude backend moving forward and provide only minimal support for it. So you may consider it in your interest to provide an implementation of this hook in the java backend as well as the maude backend. But I won't insist on it.

**Comment by omarzd** [comment]
If I delete util.k, where do I put

```
syntax KItem ::= "#time" "(" ")" [function, klabel(#time), hook(#TIME:#time)]
```

**Comment by dwightguth** [comment]
I'm not sure I understand. It seems to be duplicated from the file time.k.

**Comment by omarzd** [comment]
Yeah you're right. I will delete util.k
I don't see how the IO server is called in tcp.k.

**Comment by dwightguth** [comment]
Take a look at the rules which include "#tcpRequest" in them.

**Comment by omarzd** [comment]
Regarding the backend, How do I implement a java hook? I just imitated code from the #external pull request that I linked in the first comment.

**Comment by dwightguth** [comment]
If you look at that pull request, you will see that it actually modifies code in the java backend as well. You can just add the function as a method on the BuiltinIOOperations class.

File: k-distribution/include/builtins/time.k (hunk 1 above)
Comment (dwightguth) [inline k-distribution/include/builtins/time.k]: can you please add the attribute "impure" to this function to indicate to the java backend that this function accesses state outside its inputs?

**Pull request closed without merging**